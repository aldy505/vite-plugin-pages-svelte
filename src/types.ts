import type { WrappedComponent } from 'svelte-spa-router';

export type ImportMode = 'sync' | 'async';
export type ImportModeResolveFn = (filepath: string) => ImportMode;

export interface Route {
  [name: string]: string | ((options: WrapOptions) => WrappedComponent);
}

export interface PreRoute {
  path: string;
  component: string | ((options: WrapOptions) => WrappedComponent);
}

interface WrapOptions {
  asyncComponent: () => Promise<string>;
}

export interface PageDirOptions {
  dir: string;
  baseRoute: string;
}

/**
 * Plugin options.
 */
interface Options {
  /**
   * Relative path to the directory to search for page components.
   * @default 'src/pages'
   */
  pagesDir: string | (string | PageDirOptions)[];
  /**
   * Valid file extensions for page components.
   * @default ['svelte']
   */
  extensions: string[];
  /**
   * List of path globs to exclude when resolving pages.
   */
  exclude: string[];
  /**
   * Import routes directly or as async components
   * @default 'async'
   */
  importMode: ImportMode | ImportModeResolveFn;
  /**
   * Sync load top level index file
   * @default true
   */
  syncIndex: boolean;
  /**
   * Extend route records
   */
  extendRoute?: (route: Route, parent: Route | undefined) => Route | void;
  /**
   * Custom generated routes
   */
  onRoutesGenerated?: (routes: Route) => Route | void | Promise<Route | void>;
  /**
   * Custom generated client code
   */
  onClientGenerated?: (clientCode: string) => string | void | Promise<string | void>;
}

export type UserOptions = Partial<Options>;

export interface ResolvedPage {
  dir: string;
  route: string;
  extension: string;
  filepath: string;
  component: string;
}

export type ResolvedPages = Map<string, ResolvedPage>;

export interface ResolvedOptions extends Options {
  /**
   * Resolves to the `root` value from Vite config.
   * @default config.root
   */
  root: string;
  /**
   * RegExp to match extensions
   */
  extensionsRE: RegExp;
  pagesDir: PageDirOptions[];
}
