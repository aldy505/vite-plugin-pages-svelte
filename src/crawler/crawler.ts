import { opendir } from 'fs/promises';
import { resolve, extname } from 'path';
import type { FileOutput } from '../types/page';
import { containsExtension } from '../utils/validate';

export async function traverse(path: string, extensions: string[], ignore: string[]): Promise<FileOutput[]> {
  const r: FileOutput[] = [];
  const dir = await opendir(path, { bufferSize: 64 });
  for await (const dirent of dir) {
    if (dirent.isFile()) {
      if (ignore.includes(dirent.name)) continue;
      if (extensions.length > 0 && !containsExtension(dirent.name, extensions)) continue;
      r.push({ path: resolve(path, dirent.name) });
    } else if (dirent.isDirectory()) {
      if (ignore.includes(dirent.name)) continue;
      r.push({
        path: resolve(path, dirent.name),
        children: await traverse(resolve(path, dirent.name), extensions, ignore),
      });
    }
  }
  return r;
}

export async function traverseDir(path: string, ignore: string[]): Promise<string[]> {
  // If it's a file, then do nothing.
  if (extname(path)) return [];

  const d: string[] = [];
  const dir = await opendir(path, { bufferSize: 64 });
  for await (const dirent of dir) {
    if (dirent.isDirectory()) {
      if (ignore.includes(dirent.name)) continue;
      d.push(dirent.name);
    }
  }
  return d;
}

export function haveChildren(files: FileOutput): boolean {
  if (files.children && Array.isArray(files.children) && files.children.length > 0) return true;
  return false;
}
