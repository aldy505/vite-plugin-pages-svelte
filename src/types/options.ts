/**
 * Plugin options.
 */
interface Options {
  /**
   * Relative path to the directory to search for page components.
   * @default 'src/pages'
   */
  pagesDir: string;
  /**
   * Valid file extensions for page components.
   * @default ['svelte']
   */
  extensions: string[];
  /**
   * List of path globs to exclude when resolving pages.
   */
  exclude: string[];
  /**
   * Sync load top level index file
   * @default true
   */
  syncIndex: boolean;
}

export type UserOptions = Partial<Options>;

export interface ResolvedOptions extends Options {
  /**
   * Resolves to the `root` value from Vite config.
   * @default config.root
   */
  root: string;
  /**
   * RegExp to match extensions
   */
  extensionsRE: RegExp;
}

export type ImportMode = 'sync' | 'async';
export type ImportModeResolveFn = (filepath: string) => ImportMode;
