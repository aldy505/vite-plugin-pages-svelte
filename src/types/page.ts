export interface PageDirOptions {
  dir: string;
  baseRoute: string;
}

export interface ResolvedPage {
  dir: string;
  route: string;
  extension: string;
  filepath: string;
  component: string;
}

export type ResolvedPages = Map<string, ResolvedPage>;

export interface FileOutput {
  path: string;
  children?: FileOutput[];
}
