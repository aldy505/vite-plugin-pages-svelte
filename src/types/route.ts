import type { Route as SvelteRoute } from 'svelte-router-spa/types/components/router';

export type Route = SvelteRoute;

export interface PreRoute {
  name: string;
  path: string;
  children?: PreRoute[];
}
