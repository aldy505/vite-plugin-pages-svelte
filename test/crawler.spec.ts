import { haveChildren, traverse } from '../src/crawler/crawler';
import { FileOutput } from '../src/types/page';

const testPagesDir = 'test/assets/pages';
const testDeepPagesDir = 'test/assets/deep-pages';

describe('Traverse test page dirs', () => {
  test('Page files', async () => {
    const result = await traverse(testPagesDir, ['svelte'], []);
    expect(result.sort()).toMatchSnapshot('traverse page files');
  });
});

describe('Traverse test deep pages dir', () => {
  test('Deep pages files', async () => {
    const result = await traverse(testDeepPagesDir, ['svelte'], []);
    expect(result.sort()).toMatchSnapshot('traverse deep pages files');
  });
});

describe('Have children', () => {
  test('Should return true', () => {
    const files: FileOutput = {
      path: '/',
      children: [{ path: '/about' }],
    };
    const result = haveChildren(files);
    expect(result).toBe(true);
  });
  test('Should return false', () => {
    const files: FileOutput = {
      path: '/',
    };
    const result = haveChildren(files);
    expect(result).toBe(false);
  });
});
