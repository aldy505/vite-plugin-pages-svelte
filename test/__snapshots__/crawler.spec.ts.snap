// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Traverse test deep pages dir Deep pages files: traverse deep pages files 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/deep-pages/foo/pages/index.svelte",
          },
        ],
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/deep-pages/foo/pages",
      },
    ],
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/deep-pages/foo",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/deep-pages/bar/pages/index.svelte",
          },
        ],
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/deep-pages/bar/pages",
      },
    ],
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/deep-pages/bar",
  },
]
`;

exports[`Traverse test page dirs Page files: traverse page files 1`] = `
Array [
  Object {
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/about.svelte",
  },
  Object {
    "children": Array [
      Object {
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/about/index.svelte",
      },
    ],
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/about",
  },
  Object {
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/index.svelte",
  },
  Object {
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[userId].svelte",
  },
  Object {
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/components.svelte",
  },
  Object {
    "children": Array [
      Object {
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[sensor]/current.svelte",
      },
      Object {
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[sensor]/[...all].svelte",
      },
    ],
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[sensor]",
  },
  Object {
    "children": Array [
      Object {
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/blog/index.svelte",
      },
      Object {
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/blog/[id].svelte",
      },
      Object {
        "children": Array [
          Object {
            "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/blog/today/index.svelte",
          },
        ],
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/blog/today",
      },
    ],
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/blog",
  },
  Object {
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/[...all].svelte",
  },
  Object {
    "children": Array [
      Object {
        "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/__test__/index.svelte",
      },
    ],
    "path": "/home/reinaldy/repository/vite-plugin-pages-svelte/test/assets/pages/__test__",
  },
]
`;
